lib/rubocop/cop/sorbet/binding_constant_without_type_alias.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+require "rubocop"
+
+module RuboCop
+  module Cop
+    module Sorbet
+      # This cop disallows binding the return value of `T.any`, `T.all`, `T.enum`
+      # to a constant directly. To bind the value, one must use `T.type_alias`.
+      #
+      # @example
+      #
+      #   # bad
+      #   FooOrBar = T.any(Foo, Bar)
+      #
+      #   # good
+      #   FooOrBar = T.type_alias { T.any(Foo, Bar) }
+      class BindingConstantWithoutTypeAlias < RuboCop::Cop::Base
+        extend AutoCorrector
+
+        MSG = "It looks like you're trying to bind a type to a constant. " \
+          "To do this, you must alias the type using `T.type_alias`."
+        WITHOUT_BLOCK_MSG = "It looks like you're using the old `T.type_alias` syntax. " \
+          "`T.type_alias` now expects a block." \
+          'Run Sorbet with the options "--autocorrect --error-white-list=5043" ' \
+          "to automatically upgrade to the new syntax."
+
+        def_node_matcher :type_alias_without_block, <<~PATTERN
+          (send
+            (const {nil? cbase} :T)
+            :type_alias
+            $_
+          )
+        PATTERN
+
+        def_node_matcher :type_alias_with_block?, <<~PATTERN
+          (block
+            (send
+              (const {nil? cbase} :T)
+            :type_alias)
+            ...
+          )
+        PATTERN
+
+        def_node_matcher :requires_type_alias?, <<~PATTERN
+          (send
+            (const {nil? cbase} :T)
+            {
+              :all
+              :any
+              :class_of
+              :nilable
+              :noreturn
+              :proc
+              :self_type
+              :untyped
+            }
+            ...
+          )
+        PATTERN
+
+        def on_casgn(node)
+          expression = node.expression
+          return if expression.nil? # multiple assignment
```

Comment:
Technically someone could do

```ruby
A, B = T.any(Some, Thing), T.all(Something, Else)
```

which we'd want to flag, and autocorrect to

```ruby
A, B = T.type_alias { T.any(Some, Thing) }, T.type_alias { T.all(Something, Else) }
```

or more likely

```ruby
A = T.type_alias { T.any(Some, Thing) }
B = T.type_alias { T.all(Something, Else) }
```

However, it seems very unlikely anyone would do actually do this, so I'm not sure it's worth the complexity to try to detect it.